File: clojure/grade-school/src/grade_school.clj
(ns grade-school)

(defn- sort-key-value [db key value]
  (assoc db key (sort value)))

(defn- or-empty [item]
  (or item []))

(defn add [db student grade]
  (assoc db grade
    (conj
      (or-empty (db grade))
      student)))

(defn grade [db num]
  (or-empty (db num)))

(defn sorted [db]
  (into
    (sorted-map)
    (reduce-kv sort-key-value {} db)))

File: ocaml/grade-school/grade_school.ml
open Core

type school = string list Int.Map.t

let empty_school = Int.Map.empty

let add name grade school =
  Map.add_multi school ~key:grade  ~data:name

let grade grade school =
  match Map.find school grade with
    | None -> []
    | Some list -> list

let sorted school =
  Map.map school ~f:(fun list -> List.sort list ~compare:String.compare)
